python/mlc_llm/bench/__main__.py

```diff
@@ -0,0 +1,228 @@
+"""MLC LLM benchmark workload classes"""
+
+import asyncio
+import random
+from typing import Optional
+
+import numpy as np
+
+import mlc_llm
+from mlc_llm.bench.backend import SUPPORTED_BACKENDS, create_backend
+from mlc_llm.bench.dataset import SUPPORTED_DATASET, create_dataset
+from mlc_llm.bench.metrics import MetricProcessor, pretty_print_report
+from mlc_llm.bench.workload import create_workload
+from mlc_llm.cli.serve import EngineConfigOverride
+from mlc_llm.serve import EngineConfig
+from mlc_llm.support import argparse, logging
+
+logging.enable_logging()
+logger = logging.getLogger(__name__)
+
+
+def _parse_mlc_engine_config(config_str: Optional[str]) -> EngineConfig:
+    if config_str is None:
+        return None
+    engine_config_override = EngineConfigOverride.from_str(config_str)
+    return EngineConfig(
+        tensor_parallel_shards=engine_config_override.tensor_parallel_shards,
+        max_num_sequence=engine_config_override.max_num_sequence,
+        max_total_sequence_length=engine_config_override.max_total_seq_length,
+        prefill_chunk_size=engine_config_override.prefill_chunk_size,
+        sliding_window_size=engine_config_override.sliding_window_size,
+        attention_sink_size=engine_config_override.attention_sink_size,
+        max_history_size=engine_config_override.max_history_size,
+        gpu_memory_utilization=engine_config_override.gpu_memory_utilization,
+        spec_draft_length=engine_config_override.spec_draft_length,
+    )
+
+
+def _launch_mlc_server(args: argparse.argparse.Namespace):
+    return mlc_llm.serve.PopenServer(
+        model=args.model,
+        mode="server",
+        model_lib=args.mlc_model_lib,
+        enable_tracing=False,
+        host=args.host,
+        port=args.port,
+        engine_config=args.mlc_engine_config,
+    )
+
+
+def main(args: argparse.argparse.Namespace):
+    """Main benchmark entrance."""
+    random.seed(args.seed)
+    np.random.seed(args.seed)
+
+    mlc_server = None
+    if args.mlc_model_lib:
+        mlc_server = _launch_mlc_server(args)
+        mlc_server.start()
+
+    dataset = create_dataset(args)
+    backend = create_backend(args)
+    requests, warmup_requests = dataset.generate_requests(
```

Comment:
warmup requests should be separately from dataset. Dataset create requests. The warmup simply picks the first K requests and do a concurrenr run, then restart with the same requests (or just exclude the first K requests).